File: py/day08.py
import os

# https://adventofcode.com/2021/day/8

with open(os.path.dirname(os.path.realpath(__file__)) + "/../input/day08-input", "r") as f:
    input = [split for split in [line.strip().split(' | ') for line in f.readlines()]]

## Sample input  # expect P1: 26; P2: 61229
# lines = [
#   'be cfbegad cbdgef fgaecd cgeb fdcge agebfd fecdb fabcd edb | fdgacbe cefdb cefbgd gcbe',
#   'edbfga begcd cbg gc gcadebf fbgde acbgfd abcde gfcbed gfec | fcgedb cgb dgebacf gc',
#   'fgaebd cg bdaec gdafb agbcfd gdcbef bgcad gfac gcb cdgabef | cg cg fdcagb cbg',
#   'fbegcd cbd adcefb dageb afcb bc aefdc ecdab fgdeca fcdbega | efabcd cedba gadfec cb',
#   'aecbfdg fbg gf bafeg dbefa fcge gcbea fcaegb dgceab fcbdga | gecf egdcabf bgf bfgea',
#   'fgeab ca afcebg bdacfeg cfaedg gcfdb baec bfadeg bafgc acf | gebdcfa ecba ca fadegcb',
#   'dbcfg fgd bdegcaf fgec aegbdf ecdfab fbedc dacgb gdcebf gf | cefg dcbef fcge gbcadfe',
#   'bdfegc cbegaf gecbf dfcage bdacg ed bedf ced adcbefg gebcd | ed bcgafe cdgba cbgef',
#   'egadfb cdbfeg cegd fecab cgb gbdefca cg fgcdab egfdb bfceg | gbdfcae bgc cg cgb',
#   'gcafb gcf dcaebfg ecagb gf abcdeg gaef cafbge fdbac fegbdc | fgae cfgab fg bagce'
# ]
# input = [split for split in [line.strip().split(' | ') for line in lines]]

part1 = 0 # count all results that contain strings of length 2,3,4,7
for outp in [i[1].strip().split(' ') for i in input]:
    part1 += sum(1 for op in outp if len(op) in [2,3,4,7])

"""
Segments:
 0: 6;  [a,b,c,e,f,g]
 1: 2;  [c,f]
 2: 5;  [a,c,d,e,g]
 3: 5;  [a,c,d,f,g]
 4: 4;  [b,c,d,f]
 5: 5;  [a,b,d,f,g]
 6: 6;  [a,b,d,e,f,g]
 7: 3;  [a,c,f]
 8: 7;  [a,b,c,d,e,f,g]
 9: 6;  [a,b,c,d,f,g]

in lhs
  find el of length 2 (ie 1, segments c,f)
    el[i] -> c|f

  find el of length 3 (ie 7, segments a,c,f)
    el[i] -> if already found then c|f else a

  find el of length 4 (ie 4, segments b,c,d,f)
    el[i] -> if already found then c|f else b|d

  find el of length 7 (ie 8, all segments)
    el[i] -> if already found then a|[c|f]|[b|d] else e|g

A quick sort() and print() of the input shows that each line has exactly
one entry each of 2, 3, 4 and 7 characters and 3 entries each of 5 and 6 characters.
Use this fact to easily get and analyse entries of length 2,3,4,7.
"""

part2 = 0
for line in input:
    lhs = line[0].strip().split(' ')
    lhs.sort(key=len)
    #print(lhs)

    # 1st item == 2 chars == 1 => c|f
    scrambled = { lhs[0][0]: ['c', 'f'], lhs[0][1]: ['c', 'f'] }
    correct = { 'a': [], 'b': [], 'c': [lhs[0][0], lhs[0][1]] }

    # 2nd item == 3 chars == 7 => a|c|f - we've already 'seen' c|f so the one we haven't yet seen must be a
    for i in range(3):
        if lhs[1][i] not in scrambled:
            scrambled[lhs[1][i]] = ['a']
            correct['a'].append(lhs[1][i])

    # 3rd item == 4 chars == 4 => b|c|d|f - already seen c|f so others must be b|d
    for i in range(4):
        if lhs[2][i] not in scrambled:
            scrambled[lhs[2][i]] = ['b', 'd']
            correct['b'].append(lhs[2][i])

    # 10th item == 7 chars == 8 - already seen a|b|c|d|f so others must be e|g
    # but we don't really care

    rhs = line[1].strip().split(' ')
    numstr = ''
    for r in rhs:
        lr = len(r)
        if lr == 2:
            numstr += '1'
        elif lr == 3:
            numstr += '7'
        elif lr == 4:
            numstr += '4'
        elif lr == 7:
            numstr += '8'
        elif lr == 5:     # '2','3','5' -> (c & f) = 3; (b & d) = 5; (e & g) = 2
            if correct['c'][0] in r and correct['c'][1] in r:
                numstr += '3'
            elif correct['b'][0] in r and correct['b'][1] in r:
                numstr += '5'
            else:
                numstr += '2'
        else:             # '0','6','9' -> !(c & f) = 6; !(b & d) = 0; !(e & g) = 9
            if not (correct['c'][0] in r and correct['c'][1] in r):
                numstr += '6'
            elif not (correct['b'][0] in r and correct['b'][1] in r):
                numstr += '0'
            else:
                numstr += '9'
    #print(numstr)
    part2 += int(numstr, 10)

print("Day 8")
print("  Part 1:", part1)
print("  Part 2:", part2)

File: cpp/src/day08.cpp
#include "aocutil.h"

using namespace std;

void day08()
{
	auto input = get_input_lines(8, false);

    int part1 = 0;
    for (auto &&line : input) {
        auto rhs = split(split(line, " | ")[1], ' ');
        for (auto &&r : rhs) {
            int rl = r.length();
            if (rl == 2 || rl == 3 || rl == 4 || rl == 7) {
                part1++;
            }
        }
    }

    /* Brief analysis of the input shows that each input section of each line
       has exactly one entry each of 2, 3, 4 and 7 characters and 3 entries
       each of both 5 and 6 characters.
       Therefore
         item with 2 elements must map to '1', (c.f)
         3 el => '7' == (a,c,f)
         4 el => '4' == (b,c,d,f)
         7 el => '8' == (a,b,c,d,e,f,g)
         5 el => '2' | '3' | '5' == (a,c,d,e,g) | (a,c,d,f,g) | (a,b,d,f,g)
         6 el => '0' | '6' | '9' == (a,b,c,e,f,g) | (a,b,d,e,f,g) | (a,b,c,d,f,g)
    */
    int part2 = 0;
    for (auto &&line : input) {
        auto inout = split(line, " | ");
        auto lhs = split(inout[0], ' '), rhs = split(inout[1], ' ');

        // Sort - by length NOT alphabetically!
        sort(lhs.begin(), lhs.end(), [](const string& first, const string& second){
            return first.length() < second.length();
        });
        map<char, vector<char>> chmap{};
        map<char, vector<char>> found{};
        // 1st sorted item will have 2 chars, therefore it must be '1', and map to chars c & f
        auto li = lhs[0];
        chmap[li[0]].insert(chmap[li[0]].end(), {'c', 'f'});
        chmap[li[1]].insert(chmap[li[1]].end(), {'c', 'f'});
        found['c'].insert(found['c'].end(), {li[0], li[1]});

        // don't need to bother about the mappping for 'a' which we could derive from the 3 character item == '7'

        li = lhs[2];  // 4 ch => (b,c,d,f) with c,f already seen, therefore other two must be b,d
        for (int i = 0; i < 4; i++) {
            if (!chmap.count(li[i])) {
                chmap[li[i]].push_back('b');
                chmap[li[i]].push_back('d');
                found['b'].push_back(li[i]);
                // don't need to bother storing the 'd'
            }
        }

        string numstr = "";
        for (auto &&r : rhs) {
            switch (r.length()) {
                case 2:
                    numstr += '1';
                    break;
                case 3:
                    numstr += '7';
                    break;
                case 4:
                    numstr += '4';
                    break;
                case 5:
                    // if it contains c & f then it must be '3' (ie it contains both the choices for 'c')
                    if ((r.find(found['c'][0]) != std::string::npos) && (r.find(found['c'][1]) != std::string::npos)) {
                        numstr += '3';
                    // if it contains b & d then it must be '5' (ie it contains both the choices for 'b')
                    } else if (r.find(found['b'][0]) != std::string::npos && r.find(found['b'][1]) != std::string::npos) {
                        numstr += '5';
                    } else {
                        numstr += '2';
                    }
                    break;
                case 6:  // 0,6,9 -> opposite of cases for 2,3,5
                    if (!((r.find(found['c'][0]) != std::string::npos) && (r.find(found['c'][1]) != std::string::npos))) {
                        numstr += '6';
                    } else if (!(r.find(found['b'][0]) != std::string::npos && r.find(found['b'][1]) != std::string::npos)) {
                        numstr += '0';
                    } else {
                        numstr += '9';
                    }
                    break;
                case 7:
                    numstr += '8';
                    break;
                default:
                    cout << "Cannot handle output item " << r << endl;
                    return;
            }
        }
        // cout << numstr << endl;
        part2 += stoi(numstr);
    }

	printDayResults(8, part1, part2);
}
